future-apps/staking/contracts/Staking.sol

```diff
@@ -0,0 +1,221 @@
+    unstakeScript = _unstakeScript;
+    lockScript = _lockScript;
+
+    initialized();
+  }
+
+  function stake(uint256 amount, bytes data) authP(STAKE_ROLE, arr(amount)) public {
+    stakeFor(msg.sender, amount, data);
+  }
+
+  function stakeFor(address acct, uint256 amount, bytes data) authP(STAKE_ROLE, arr(amount)) public {
+    // From needs to be msg.sender to avoid token stealing by front-running
+    require(stakingToken.transferFrom(msg.sender, this, amount));
+    processStake(acct, amount, data);
+
+    if (stakeScript.length > 0) {
+      runScript(stakeScript, data, new address[](0));
+    }
+  }
+
+  function processStake(address acct, uint256 amount, bytes data) internal {
+    accounts[acct].amount += amount; // overflow check in token contract
+    assert(accounts[acct].amount >= amount);
+
+    Staked(acct, amount, totalStakedFor(acct), data);
+  }
+
+  function unstake(uint256 amount, bytes data) authP(UNSTAKE_ROLE, arr(amount)) checkUnlocked(amount) public {
+    require(accounts[msg.sender].amount >= amount);
+    accounts[msg.sender].amount -= amount;
+
+    require(stakingToken.transfer(msg.sender, amount));
+
+    Unstaked(msg.sender, amount, totalStakedFor(msg.sender), data);
+
+    if (unstakeScript.length > 0) {
+      runScript(unstakeScript, data, new address[](0));
+    }
+  }
+
+  function lock(
+    uint256 amount,
+    uint8 lockUnit,
+    uint64 lockEnds,
+    address unlocker,
+    bytes32 metadata,
+    bytes data
+  )
+    authP(LOCK_ROLE, arr(amount, uint256(lockUnit), uint256(lockEnds)))
+    checkUnlocked(amount)
+    public
+  {
+    Lock memory newLock = Lock(amount, Timespan(lockEnds, TimeUnit(lockUnit)), unlocker, metadata);
+    uint256 lockId = accounts[msg.sender].locks.push(newLock) - 1;
+
+    Locked(msg.sender, lockId, metadata);
+
+    if (lockScript.length > 0) {
+      runScript(lockScript, data, new address[](0));
+    }
+  }
+
+  function stakeAndLock(
+    uint256 amount,
+    uint8 lockUnit,
+    uint64 lockEnds,
+    address unlocker,
+    bytes32 metadata,
+    bytes stakeData,
+    bytes lockData
+  )
+    authP(STAKE_ROLE, arr(amount))
+    authP(LOCK_ROLE, arr(amount, uint256(lockUnit), uint256(lockEnds)))
+    public
+  {
+    stake(amount, stakeData);
+    lock(amount, lockUnit, lockEnds, unlocker, metadata, lockData);
+  }
+
+  function unlockAll(address acct) external {
+    while (accounts[acct].locks.length > 0) {
+      if (canUnlock(acct, 0)) {
+        unlock(acct, 0);
+      }
+    }
+  }
+
+  function unlock(address acct, uint256 lockId) public {
+    require(canUnlock(acct, lockId));
+
+    uint256 lastAccountLock = accounts[acct].locks.length - 1;
+    if (lockId < lastAccountLock) {
+      accounts[acct].locks[lockId] = accounts[acct].locks[lastAccountLock];
+    }
+
+    accounts[acct].locks.length -= 1;
+
+    Unlocked(acct, msg.sender, lockId);
+  }
+
+  function moveTokens(address from, address to, uint256 amount) authP(GOD_ROLE, arr(from, to, amount)) external {
```

Comment:
Shouldn't we check if moved tokens are unlocked? Otherwise an account could end up having more locked tokens than staked.